scripts/patch_release/patch.py

```diff
@@ -0,0 +1,358 @@
+"""
+Patch releases occur when samples need to be retracted due to
+patients withdrawing consent.
+
+Patches should always occur on the latest consortium release for
+the specific public release. Due to the GENIE retraction policy,
+it is best to retract data on the 3rd consortium release of the
+subsequent release series.
+"""
+import argparse
+import os
+import shutil
+import tempfile
+
+import pandas as pd
+import synapseclient
+
+from genie import process_functions
+from genie import create_case_lists
+from genie import dashboard_table_updater
+
+
+# Run time functions
+def revise_meta_file(meta_file_path: str, old_version: str, new_version: str) -> None:
+    """
+    Replaces the old version with the new version in the meta file.
+
+    Args:
+        meta_file_path (str): The path to the meta file.
+        old_version (str): The old version to be replaced.
+        new_version (str): The new version to replace the old version.
+
+    Returns:
+        None
+    """
+    with open(meta_file_path, "r") as meta:
+        meta_text = meta.read()
+    with open(meta_file_path, "w") as meta:
+        meta_text = meta_text.replace(old_version, new_version)
+        meta.write(meta_text)
+
+
+def store_file(
+    syn: synapseclient.Synapse, new_path: str, new_release_synid: str, release_name: str
+) -> None:
+    """
+    Stores a file into Synapse.
+
+    Args:
+        syn (synapseclient.Synapse): The Synapse client object.
+        new_path (str): The path to the file to be stored.
+        new_release_synid (str): The Synapse ID of the release folder where the file will be stored.
+        release_name (str): The name of the release.
+
+    Returns:
+        None
+    """
+    ent_name = os.path.basename(new_path.replace(f"_{release_name}", ""))
+    new_ent = synapseclient.File(new_path, name=ent_name, parentId=new_release_synid)
+    syn.store(new_ent)
+
+
+def patch_release_workflow(
+    release_synid: str, new_release_synid: str, retracted_sample_synid: str
+):
+    """
+    These need to be modified per retraction.
+    The release_synid, new_release_synid, and retracted_sample_synid
+    variables need to be changed to reflect different Synapse ids per release.
+    """
+    syn = synapseclient.login()
+
+    remove_centers = []
+    remove_seqassays = []
+    # 11 release series
+    release_synid = ""  # Fill in synapse id here
+    old_release = syn.get(release_synid).name
+    # The new release folder MUST be created on synapse first.
+    new_release_synid = ""  # Fill in synapse id here
+    new_release = syn.get(new_release_synid).name
+    # samples to retract. Example: syn27734047 (12.3 consortium)
+    # If you are creating the 11.1-public patch, you will be using the
+    # 12.3-consortium samples to retract file
+    # Synapse id configurations
+    retracted_sample_synid = ""
+    # Data base mapping synid
+    database_mapping_synid = "syn10967259"
```

Comment:
Seeing this hardcoded made me wonder .. do we forsee needing to run staging for a patch release? For comparisons mainly to see if what we expect to retract went through? Just curious how we do it now

Reply:
Ya... The retraction code can be improved, but basically right now, you would have to create a synapse folder for where the retracted files lives.  So it's _ok_ to be hardcoded.

In the future, when the retraction is "retract for 13.0 series" and it automatically finds all these synapse ids, that would require non-hardcoded synapse ids here.

Reply:
Ah I see they are in a separate folder. Good to know. 

Yeah I'm fine keeping it as is since we only do it twice a year. 

I think if we forsee updating this code a few times a year in the future, it would be nice to have it run-able on like the test synapse project for example (or staging if it's easier) prior to just running it on production.